### PR TITLE
Handle `Uncancelable` and `OnCancel` in `syncStep` interpreter in 3.x

### PR DESCRIPTION
**Pull request by armanbilge**
Forward merge of https://github.com/typelevel/cats-effect/pull/3064.

### CODE REVIEW

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 1]
```diff
@@ -1961,6 +1961,17 @@ private object SyncStep {
...
+        case IO.Uncancelable(body, _)
+            if G.isInstanceOf[MonadCancel.Uncancelable[G, Throwable] @unchecked] =>
+          val ioa = body(new Poll[IO] {
+            def apply[C](ioc: IO[C]): IO[C] = ioc
+          })
+          interpret(ioa, limit)
+
+        case IO.OnCancel(ioa, _)
+            if G.isInstanceOf[MonadCancel.Uncancelable[G, Throwable] @unchecked] =>
+          interpret(ioa, limit)
```
Comment (armanbilge) [inline core/shared/src/main/scala/cats/effect/IO.scala:1973]: So this is kind of cheating, but I'm not sure if we can do better.

For `OnCancel`, what happens if we need to bail out the `finalizer`? Then we would need to return a `Left(...)` but I don't see how that's possible if we're in the process of being canceled.

I haven't thought as deeply about `Uncancelable`, but the mental gymnastics of implementing `Poll[IO]` in terms of `Poll[G]` were enough to put me off 😂

File: core/shared/src/main/scala/cats/effect/IO.scala [hunk 2]
```diff
@@ -1961,6 +1961,17 @@ private object SyncStep {
             }
             .handleErrorWith(t => interpret(f(t), limit - 1))
 
+        case IO.Uncancelable(body, _)
+            if G.isInstanceOf[MonadCancel.Uncancelable[G, Throwable] @unchecked] =>
```
Comment (djspiewak) [inline core/shared/src/main/scala/cats/effect/IO.scala:1965]: Okay, so finally looped around to this. The correct way to do something like this is via `rootCancelScope`:

```suggestion
        case IO.Uncancelable(body, _) if G.rootCancelScope == CancelScope.Uncancelable =>
```

**Review by djspiewak — CHANGES_REQUESTED** [review]

File: core/shared/src/main/scala/cats/effect/IO.scala (hunk 2 above)
Comment (armanbilge) [reply core/shared/src/main/scala/cats/effect/IO.scala:1965]: Ah, right! Forgot about this.

Comment (djspiewak) [reply core/shared/src/main/scala/cats/effect/IO.scala:1965]: I already made the change locally :-)

**Pull request merged**